Synchronized on database when unlock tables

nivolg opened the pull request:
I found a problem with table locking under heavy load. The error "Timeout trying to lock table T" happened when I try to drop a table. I made some tests in debug mode and found a cause of the problem: after multiple concurrent table reading variable  RegularTable.lockSharedSessions becomes corrupted (size = 1 but no elements inside) so every next drop attempt fails. In Session.unlockAll there is no synchronizing  on database section as it is in other similar methods.

I use h2 with  MULTI_THREADED=1 and MV_STORE=FALSE properties because we only read from tables after they were created. 

Also I found a workaround - set autocommit=false for select queries. In this case session invoke unlockReadLocks methods where  synchronized (database) is present.

nivolg committed:
Synchronized on database when unlock tables

grandinj commented:
I think you should just move the code in RegularTable#unlock so that lockSharedSessions is only modified when synchronized on database.
Because we don't seem to be synchronizing on Database inside Session when we are touching the Session#locks object.

Unless you can reproduce this when in MV_STORE=TRUE mode?
In which case I would agree that the problem might lie in the Session class.

grandinj commented:
Also,if possible, a unit test would be really really nice :-)

svladykin commented:
Guys,

First of all I don't like that we are adding another global synchronization point to H2 code. We have to avoid it as much as possible if we want MULTI_THREADED=1 scale well.

Also it is wrong to modify `Session` here, because I see that  in `RegularTable` (which is used when MV_STORE=FALSE) we explicitly synchronize on `database` in methods `lock(Session, boolean, boolean)` and `unlock(Session)`, but in `MVTable` (MV_STORE=TRUE) we synchronized on `getLockSyncObject()` which must not be `database` with MULTI_THREADED=1.

I guess it is better to start with a failing test, run it in all the possible modes and then decide how to fix the issue.

nivolg committed:
Synchronized on database when remove from lockSharedSessions, add test

nivolg commented:
Hi again, I added a test which fails without my fix and removed synchronization from session simultaneously with extending of synchronizing block in RegularTable.unlock. A such fix is more correct, I agree with Noel.
Also I wasn't able to reproduce the bug with enabled MV_STORE.

svladykin commented:
This fix looks much better to me.
